Add elementwise_affine for layer_norm_op

xiaomengy opened the pull request:
Summary: Add elementwise_affine for layer_norm_op

Differential Revision: D15075454

houseroad commented on caffe2/operators/layer_norm_op.cc:
```diff
@@ -174,10 +191,16 @@ to the end.)
         0,
         "input",
         "Input tensor which layer normalization will be applied to")
+    .Input(1, "gamma", "scale tensor for elementwise_affine")
```
specify expected shape?

houseroad commented on caffe2/operators/layer_norm_op.cc:
```diff
@@ -174,10 +191,16 @@ to the end.)
         "(float) default to 0.001. Small value to be added to the stdev when"
         " dividing out by that value. This prevents division by zero.")
+    .Arg(
+        "elementwise_affine",
```
This seems redundant, since we can determine whether we should do the elementwise_affine based on the input size.

smessmer replied:
I think being explicit is always better and more readable. Relying on input size can be confusing. But we should probably add asserts to make sure we're in one of the valid states.

BIT-silence replied:
It will help to determine whether to compute gradients for gamma and beta.

houseroad commented on caffe2/python/operator_test/layer_norm_op_test.py:
```diff
@@ -6,28 +6,35 @@
 
 
-def _layer_norm_ref(axis, epsilon, X):
+def _layer_norm_ref1(axis, epsilon, X):
```
nit: _layer_norm_ref

houseroad commented on caffe2/python/operator_test/layer_norm_op_test.py:
```diff
@@ -6,28 +6,35 @@
+    return (Y, mean, std)
+
+
+def _layer_norm_ref2(axis, epsilon, X, gamma, beta):
```
nit: _layer_norm_with_affine_ref?

houseroad commented on caffe2/utils/math/reduce.cc:
```diff
@@ -663,6 +663,7 @@ DELEGATE_REDUCE_FUNCTION(double, ReduceL2, 0.0, true)
     MomentsImpl<T>(ndim, X_dims, Y_dims, X, mean, var, context); \
   }
 CAFFE2_SPECIALIZED_MOMENTS(float)
+CAFFE2_SPECIALIZED_MOMENTS(double)
```
Is this necessary?

BIT-silence replied:
Just in case some use case in PyTorch that the model is training in double.

houseroad reviewed (commented):
Overall, looks good. Please check inline comments.

smessmer commented on caffe2/core/operator_c10wrapper.h:
```diff
@@ -103,15 +108,14 @@ class C10OperatorWrapper final : public Operator<Context> {
-          input_tensor_index,
-          ") in the c10 operator schema.");
     }
+    AT_ASSERTM(
```
thanks. awesome catch.

houseroad reviewed (approved):
Looks good to me.

xiaomengy committed:
Add elementwise_affine for layer_norm_op (#19713)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/19713

Add elementwise_affine for layer_norm_op

Reviewed By: houseroad

Differential Revision: D15075454

fbshipit-source-id: 91a04ee55cebf81b253dca69a8d08da9b2f8b0fb
